### ISSUE TITLE
CanConnect() and other methods throw NullReferenceException when run as an Azure container, but works fine on development pc.

### ISSUE THREAD

**Issue by wrkettlitz** [dotnet/efcore#18760]
``` C#
2019-11-05T14:16:55.759289586Z [40m[37mdbug[39m[22m[49m: Microsoft.EntityFrameworkCore.Infrastructure[10401]
2019-11-05T14:16:55.759373088Z       An 'IServiceProvider' was created for internal use by Entity Framework.
2019-11-05T14:16:56.593097615Z [40m[32minfo[39m[22m[49m: Secret.Methods.InitializeDB[0]
2019-11-05T14:16:56.593150016Z       Testing DbConnection
2019-11-05T14:17:01.811427329Z Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
2019-11-05T14:17:01.811468330Z    at Microsoft.EntityFrameworkCore.Diagnostics.CoreLoggerExtensions.RequiredAttributeOnCollection(IDiagnosticsLogger`1 diagnostics, INavigation navigation)
2019-11-05T14:17:01.811479831Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.RequiredNavigationAttributeConvention.ProcessNavigationAdded(IConventionRelationshipBuilder relationshipBuilder, IConventionNavigation navigation, RequiredAttribute attribute, IConventionContext`1 context)
2019-11-05T14:17:01.811489631Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.NavigationAttributeConventionBase`1.ProcessNavigationAdded(IConventionRelationshipBuilder relationshipBuilder, IConventionNavigation navigation, IConventionContext`1 context)
2019-11-05T14:17:01.811498831Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnNavigationAdded(IConventionRelationshipBuilder relationshipBuilder, IConventionNavigation navigation)
2019-11-05T14:17:01.811508131Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.RunVisitor.VisitOnNavigationAdded(OnNavigationAddedNode node)
2019-11-05T14:17:01.811517032Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnNavigationAddedNode.Accept(ConventionVisitor visitor)
2019-11-05T14:17:01.811527132Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionVisitor.Visit(ConventionNode node)
2019-11-05T14:17:01.811535832Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionVisitor.VisitConventionScope(ConventionScope node)
2019-11-05T14:17:01.811544432Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Run()
2019-11-05T14:17:01.811561533Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Dispose()
2019-11-05T14:17:01.811570233Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnModelInitialized(IConventionModelBuilder modelBuilder)
2019-11-05T14:17:01.811580133Z    at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnModelInitialized(IConventionModelBuilder modelBuilder)
2019-11-05T14:17:01.811588933Z    at Microsoft.EntityFrameworkCore.Metadata.Internal.Model..ctor(ConventionSet conventions)
2019-11-05T14:17:01.811597034Z    at Microsoft.EntityFrameworkCore.ModelBuilder..ctor(ConventionSet conventions)
2019-11-05T14:17:01.811604934Z    at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder)
2019-11-05T14:17:01.811613234Z    at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.<>c__DisplayClass5_0.<GetModel>b__1()
2019-11-05T14:17:01.811621934Z    at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
2019-11-05T14:17:01.811629635Z    at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
2019-11-05T14:17:01.811637735Z    at System.Lazy`1.CreateValue()
2019-11-05T14:17:01.811645535Z    at System.Lazy`1.get_Value()
2019-11-05T14:17:01.811653635Z    at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder)
2019-11-05T14:17:01.811661935Z    at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
2019-11-05T14:17:01.811669936Z    at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
2019-11-05T14:17:01.811677936Z    at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_3(IServiceProvider p)
2019-11-05T14:17:01.811686736Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
2019-11-05T14:17:01.811695236Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
2019-11-05T14:17:01.811703636Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
2019-11-05T14:17:01.811712137Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
2019-11-05T14:17:01.811740237Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
2019-11-05T14:17:01.811750738Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
2019-11-05T14:17:01.811765138Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
2019-11-05T14:17:01.811773938Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
2019-11-05T14:17:01.811782739Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
2019-11-05T14:17:01.811790939Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
2019-11-05T14:17:01.811799039Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
2019-11-05T14:17:01.811807339Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
2019-11-05T14:17:01.811816039Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
2019-11-05T14:17:01.811824340Z    at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
2019-11-05T14:17:01.811832540Z    at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
2019-11-05T14:17:01.811840740Z    at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
2019-11-05T14:17:01.811848840Z    at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
2019-11-05T14:17:01.811856740Z    at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
2019-11-05T14:17:01.811864441Z    at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
2019-11-05T14:17:01.811873041Z    at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
2019-11-05T14:17:01.811882341Z    at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
2019-11-05T14:17:01.811890541Z    at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_Dependencies()
2019-11-05T14:17:01.811898442Z    at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.CanConnect()
2019-11-05T14:17:01.811906542Z    at secret.Methods.InitializeDB.Initialize(IServiceProvider serviceProvider) in /app/Iris Online Mobiel/Methods/InitializeDB.cs:line 35
2019-11-05T14:17:01.811914742Z    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
2019-11-05T14:17:01.811922942Z    at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
2019-11-05T14:17:01.811930842Z    at System.Threading.ThreadPoolWorkQueue.Dispatch()
2019-11-05T14:17:01.830478228Z    at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

### Further technical details

EF Core version:
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.0
Operating system: Docker Container Ubuntu
IDE: Visual Studio 2019 latest preview.

**Comment by wrkettlitz** [comment]
[Looks like this is what fails? I am completly clueless why this happens. I appreciate any help you can offer.](https://imgur.com/a/C7qc2cn)

**Comment by wrkettlitz** [comment]
Okay, looks like it crashed because i had an Requiredattribute on a list, I will still keep this issue open because i dont think it is supposed to throw a NullReferenceError, besides, i still dont understand why the error happend in azure but not on the developer pc. So I believe these are 2 bugs.

**Comment by ajcvickers** [comment]
Note for triage: key piece here is that the relationship needs to be unidirectional. This results in a null INavigation being passed to the logger.

/cc @AndriySvyryd 

```C#
public class Blog
{
    public int Id { get; set; }
    public int NotId { get; set; }
    
    [Required]
    public List<Post> Posts { get; set; }
}

public class Post
{
    public int Id { get; set; }
    //public Blog Blog { get; set; }
}

public class BloggingContext : DbContext
{
    private readonly ILoggerFactory Logger 
        = LoggerFactory.Create(c => c.AddConsole().SetMinimumLevel(LogLevel.Debug));

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder
            .UseLoggerFactory(Logger)
            .UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Test;ConnectRetryCount=0");
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Blog>();
    }
}

public class Program
{
    public static async Task Main()
    {
        using (var context = new BloggingContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();
        }
    }
}
```

**Comment by AndriySvyryd** [comment]
@wrkettlitz [Required] on a collection navigation doesn't affect configuration and can be removed. You see a different behavior on Azure container because it has different logging configuration.

**Comment by nathan-c** [comment]
If I run my service without ASPNETCORE_ENVIRONMENT=Development set then I also see this problem. Any hints for what logging config i need to change to get this working again?

**Comment by nathan-c** [comment]
if anyone else has same problem I was able to avoid the exception by changing my logging config from 
``` json
  "Logging": {
    "LogLevel": {
      "Default": "debug",
      "System": "Information",
      "Microsoft": "Information"
    }
  }
```
to
``` json
  "Logging": {
    "LogLevel": {
      "Default": "debug"
    }
  }
```

**Comment by AndriySvyryd** [comment]
Fixed in e54cf7ba711a0f58b76e3598e994854e1a284c96

**Comment by mocsharp** [comment]
Will this be fixed in the 3.1 branch? Thanks.
